### PR TITLE
Fix more Unicode bugs

### PR DESCRIPTION
**Pull request by fmeum**
* Use Latin-1 in many native file write rules for consistency with the internal encoding.
* Use Latin-1 for the resolved repository file and the JSON profile.
* Fix `unused_input_list` handling of non-ASCII characters in file names.
* Flip the `legacy_utf8` parameter of `repository_ctx.file` to `False` and make it a no-op. With the previous default, any non-ASCII characters would be written out as double encoded UTF-8, which is not a useful choice.
* Change `repository_ctx.template` to operate on raw bytes for consistency with `repository_ctx.read` and to fix substitution with non-ASCII keys/values.
* Move some usages of `UTF_8` closer to their usage site to clarify why they are correct.
* Fixes parsing of dependency files with Unicode character contents (`/showIncludes` and `.d` files)

### CODE REVIEW

**Comment by fmeum** [comment]
@tjgq This is stacked on #24010, but the last commit is new and ready for review.

**Comment by tjgq** [comment]
@fmeum Mind rebasing before I review?

**Comment by fmeum** [comment]
@bazel-io fork 8.0.0

File: src/test/shell/bazel/starlark_repository_test.sh [hunk 1]
```diff
@@ -470,6 +478,84 @@ EOF
   expect_log "PWD=$repo2 TOTO=titi"
 }
 
+function test_starlark_repository_unicode() {
+  setup_starlark_repository
+
+  if "$is_windows"; then
+    # äöüÄÖÜß in UTF8
```
Comment (tjgq) [inline src/test/shell/bazel/starlark_repository_test.sh:485]: Nit: "UTF-8" (also below)

File: src/test/shell/integration/starlark_dependency_pruning_test.sh [hunk 2]
```diff
@@ -242,6 +250,43 @@ function test_dependency_pruning_scenario() {
   check_unused_content "pkg/c.input"
 }
 
+function test_dependency_pruning_scenario_unicode() {
+  if "$is_windows"; then
+    # äöüÄÖÜß in UTF8
```
Comment (tjgq) [inline src/test/shell/integration/starlark_dependency_pruning_test.sh:255]: Nit: "UTF-8" (also below)

**Review by tjgq — CHANGES_REQUESTED** [review]

**Comment by tjgq** [comment]
> @tjgq Thanks for the review, it helped me find and fix more issues.
> 
> I noticed that there are a bunch of `FileSystemUtils` methods that read and write files in different encodings. I could clean that up in a follow-up PR or as part of this PR, as you prefer.

I'd prefer a separate PR, thanks!

**Comment by fmeum** [comment]
> @fmeum it looks like there's an issue with .netrc parsing? (see failed tests)

Should be fixed now

**Comment by fmeum** [comment]
@meteorcloudy Compiling a `cc_library` that depends on a header file with non-ASCII characters in its name results in this kind of error:

```
ERROR: C:/b/absmfft5/execroot/_main/_tmp/925c5432c7427d85ec55d631e57c838b/workspace/pkg/BUILD:5:10: Compiling pkg/bin.cc failed: undeclared inclusion(s) in rule '//pkg:bin':
this rule is missing dependency declarations for the following files included by 'pkg/bin.cc':
  'pkg/„”Ž™šá??.h'
```

The raw bytes of the file name are `112, 107, 103, 47, -124, -108, -127, -114, -103, -102, -31, 63, 63, 46, 104`, which indicates that the output of MSVC is in the OEM codepage 437. This is hinted at [here](https://learn.microsoft.com/en-gb/cpp/build/reference/unicode-support-in-the-compiler-and-linker?view=msvc-170#output). Do you know of a way to force `cl.exe` to output Unicode when it's not running in a console (or, alternatively, to have it think it's running in a console)?

**Comment by meteorcloudy** [comment]
Maybe this one https://learn.microsoft.com/en-gb/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8?view=msvc-170?

**Comment by fmeum** [comment]
> Maybe this one https://learn.microsoft.com/en-gb/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8?view=msvc-170?

That didn't help :-(

**Comment by fmeum** [comment]
I have given up on getting `cl.exe` to work. Users could presumably still do this themselves by switching their system code page to UTF-8.

**Comment by fmeum** [comment]
@tjgq Rebased onto master.

**Comment by fmeum** [comment]
@bazel-io fork 8.0.0

**Comment by iancha1992** [comment]
@fmeum can you please resolve the conflicts? Thanks!

**Comment by fmeum** [comment]
@iancha1992 Rebased

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 8.0.0 RC3](https://github.com/bazelbuild/bazel/releases/tag/8.0.0rc3). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=8.0.0rc3`. Thanks!

**Comment by meteorcloudy** [comment]
Looks like this is again rolled-back in https://github.com/bazelbuild/bazel/commit/e27ab916aaa6b12f2be4e463a54ef49965cc04a7

@tjgq @katre Can you provide some details on the issue?

**Comment by katre** [comment]
@tjgq has a better understanding of the cause than I do (I just noticed the failures and initiated the rollback).

**Comment by tjgq** [comment]
See b/381060195 for some details. I'm fairly confident that the root cause was in Google code, not Bazel code, and I expect the fix to be very localized. So, absent evidence that it breaks anything externally, I'd recommend sticking with the cherry-pick for now.

**Pull request closed without merging**